### DISCUSSION TITLE
R2-Kotlin:  play audio Epub  content using local server

### DISCUSSION THREAD

**Discussion by nareshsvn** [readium/kotlin-toolkit#46] [Q&A]
In R1, We used to play the audio epub using exo-player(native player) by starting EpubServer.
Start Server:
EpubServer mServer = new EpubServer(EpubServer.HTTP_HOST, EpubServer.HTTP_PORT,
                        mContainer.getDefaultPackage(), false, null, getApplicationContext());
                mServer.setAudioPlayingStatus();
                mServer.startServer();

After starting the server, we used to supply the url( mContainer.getDefaultPackage().getManifestItem(position).getHref()) to the exo-player. now exo-player plays locally. (Epub server acts as a local server and exo-payer will communicate using url "http://127.0.0.1:8080/track1")              

In R2 is there a way we can achieve the same behavior(we need to use a native player to give more controls)?

**Comment by mickael-menu** [comment]
In R2 ([at the moment](https://github.com/readium/r2-testapp-kotlin/issues/297)) resources are also served through an HTTP server, so you can also use it to provide URLs to ExoPlayer.

In the Readium Web Publication models used in R2, `readingOrder` will contain the `Link` to the EPUB chapters, while `resources` should contain CSS, JavaScript and also audio files. So you can use the following to get the list of audio files referenced in the EPUB:

```kotlin
val audioLinks = publication.resources.filter { it.mediaType.isAudio }
```

Then you can concatenate the base URL used to serve the publication on the server when adding the publication to the link's href to get the full URL:

```kotlin
val baseUrl = Publication.localBaseUrlOf(publicationFilename, server.port)
val url = baseUrl + audioLinks[0].href
```

**Reply by mickael-menu** [reply]
However, my preferred approach would be to use ExoPlayer's [DataSource](https://exoplayer.dev/media-sources.html) to implement reading directly the content of a publication using our `Resource` API. This doesn't require an HTTP server.

I implemented it in this WIP [Audiobook navigator PR](https://github.com/readium/r2-navigator-kotlin/pull/160).

Take a look in particular at `PublicationDataSource` and how it's used in the PR:

https://github.com/readium/r2-navigator-kotlin/pull/160/files#diff-50505037106ff838f537dab8e101b8f922a51e4eff71ccc87f9857e7d8228f7e
